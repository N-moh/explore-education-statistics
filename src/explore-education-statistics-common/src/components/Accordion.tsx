import useMounted from '@common/hooks/useMounted';
import isComponentType from '@common/lib/type-guards/components/isComponentType';
import classNames from 'classnames';
import React, {
  cloneElement, ComponentType,
  ReactComponentElement,
  ReactNode,
  useEffect,
  useRef, WeakValidationMap,
} from 'react';
import {useImmer} from 'use-immer';
import styles from './Accordion.module.scss';
import {
  accordionSectionClasses,
  AccordionSectionProps,
} from './AccordionSection';

export interface AccordionProps {
  children: ReactNode;
  id: string;
  onToggleAll?: (open: boolean) => void;
  onToggle?: (accordionSection: { id: string; title: string }) => void;
}


export function generateIdList(count: number) {
  return new Array(count).fill('content-section-').map((id, n) => id + (n + 1));
}

const Accordion = ({children, id, onToggleAll, onToggle}: AccordionProps) => {

  const ref = useRef<HTMLDivElement>(null);

  const [openSections, updateOpenSections] = useImmer<boolean[]>([]);

  const sections = React.Children.toArray(children) as ReactComponentElement<ComponentType<AccordionSectionProps>>[];

  const getSectionIds = (
    sectionProps: AccordionSectionProps,
    index: number,
  ) => {
    return {
      contentId: sectionProps.contentId || `${id}-${index + 1}-content`,
      headingId: sectionProps.headingId || `${id}-${index + 1}-heading`,
    };
  };

  useEffect(() => {
    updateOpenSections(() =>
      sections.map(section => {
        return section.props.open || false;
      }),
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [children, updateOpenSections]);

  const {isMounted} = useMounted(() => {
    const goToHash = () => {
      if (ref.current && window.location.hash) {
        let locationHashEl: HTMLElement | null = null;

        try {
          locationHashEl = ref.current.querySelector(window.location.hash);
        } catch (_) {
          return;
        }

        if (locationHashEl) {
          const sectionEl = locationHashEl.closest(
            `.${accordionSectionClasses.section}`,
          );

          if (sectionEl) {
            const contentEl = sectionEl.querySelector(
              `.${accordionSectionClasses.sectionContent}`,
            );

            if (contentEl) {
              updateOpenSections(draft => {
                const openIndex = sections.findIndex((section, index) => {
                  const {contentId, headingId} = getSectionIds(
                    section.props,
                    index,
                  );
                  const hashId = contentEl.id;
                  return hashId === contentId || hashId === headingId;
                });

                if (openIndex > -1) {
                  draft[openIndex] = true;
                }
              });

              setTimeout(
                () =>
                  (locationHashEl as HTMLElement).scrollIntoView({
                    block: 'start',
                  }),
                100,
              );
            }
          }
        }
      }
    };

    goToHash();
    window.addEventListener('hashchange', goToHash);

    return () => window.removeEventListener('hashchange', goToHash);
  });

  const isAllOpen = openSections.every(isOpen => isOpen);

  return (
    <div
      className={classNames('govuk-accordion', styles.accordionPrint)}
      id={id}
      ref={ref}
      role="none"
      data-module="govuk-accordion"
    >
      {isMounted && (
        <div className="govuk-accordion__controls">
          <button
            aria-expanded={isAllOpen}
            type="button"
            className="govuk-accordion__open-all"
            onClick={() => {
              updateOpenSections(draft => {
                return draft.map(() => !isAllOpen);
              });

              if (onToggleAll) {
                onToggleAll(!isAllOpen);
              }
            }}
          >
            {isAllOpen ? 'Close all ' : 'Open all '}
            <span className="govuk-visually-hidden">sections</span>
          </button>
        </div>
      )}

      {sections.map((section, index) => {
        const {headingId, contentId} = getSectionIds(section.props, index);

        const isSectionOpen = isAllOpen || openSections[index];

        return cloneElement<AccordionSectionProps>(section, {
          key: headingId,
          headingId,
          contentId,
          open: isSectionOpen,
          onToggle(isOpen) {
            updateOpenSections(draft => {
              draft[index] = isOpen;
            });

            if (onToggle && isOpen) {
              onToggle({
                id: headingId,
                title: section.props.heading,
              });
            }

            if (section.props.onToggle) {
              section.props.onToggle(isOpen);
            }
          },
        });
      })}

    </div>
  );
};

export default Accordion;
