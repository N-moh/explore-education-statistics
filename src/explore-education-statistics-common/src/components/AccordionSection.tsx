import findAllParents from '@common/lib/dom/findAllParents';
import classNames from 'classnames';
import React, { createElement, ReactNode } from 'react';
import GoToTopLink from './GoToTopLink';

export interface AccordionSectionProps {
  caption?: string;
  children: ReactNode;
  className?: string;
  contentId?: string;
  goToTop?: boolean;
  heading: string;
  headingId?: string;
  // Only for accessibility/semantic markup,
  // does not change the actual styling
  headingTag?: 'h2' | 'h3' | 'h4';
  id?: string;
  open?: boolean;
  onToggle?: (open: boolean) => void;
}

export const classes = {
  section: 'govuk-accordion__section',
  sectionButton: 'govuk-accordion__section-button',
  sectionContent: 'govuk-accordion__section-content',
  expanded: 'govuk-accordion__section--expanded',
};

const AccordionSection = ({
  caption,
  className,
  children,
  contentId,
  goToTop = true,
  heading,
  headingId,
  headingTag = 'h2',
  open = false,
  onToggle,
}: AccordionSectionProps) => {
  return (
    <div
      onClick={event => {
        if (onToggle) {
          onToggle(event.currentTarget.classList.contains(classes.expanded));
        }
      }}
      className={classNames(classes.section, className, {
        [classes.expanded]: open,
      })}
      role="presentation"
    >
      <div
        className="govuk-accordion__section-header"
        data-testid={`SectionHeader ${heading}`}
      >
        {createElement(
          headingTag,
          {
            className: 'govuk-accordion__section-heading',
          },
          <span className={classes.sectionButton} id={headingId}>
            {heading}
          </span>,
        )}
        {caption && (
          <span className="govuk-accordion__section-summary">{caption}</span>
        )}
      </div>

      <div
        className={classes.sectionContent}
        aria-labelledby={headingId}
        id={contentId}
      >
        {children}

        {goToTop && <GoToTopLink />}
      </div>
    </div>
  );
};

export default AccordionSection;

export const openAllParentAccordionSections = (target: HTMLElement) => {
  const sections = findAllParents(target, `.${classes.section}`);

  sections.forEach(section => {
    const button = section.querySelector<HTMLElement>(
      `.${classes.sectionButton}`,
    );

    if (button && button.getAttribute('aria-expanded') === 'false') {
      button.click();
    }
  });
};
