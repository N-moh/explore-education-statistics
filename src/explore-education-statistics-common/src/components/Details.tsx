import useMounted from '@common/hooks/useMounted';
import useToggle from '@common/hooks/useToggle';
import findAllParents from '@common/lib/dom/findAllParents';
import classNames from 'classnames';
import React, { MouseEvent, ReactNode, useEffect, useRef } from 'react';

let hasNativeDetails: boolean;
let idCounter = 0;

export interface DetailsProps {
  className?: string;
  tag?: string;
  children: ReactNode;
  id?: string;
  onToggle?: (isOpen: boolean, event: MouseEvent<HTMLElement>) => void;
  analytics?: (eventGA: { category: string; action: string }) => void;
  open?: boolean;
  summary: string | ReactNode;
}

const Details = ({
  className,
  children,
  id = `details-content-${(idCounter += 1)}`,
  open = false,
  onToggle,
  analytics,
  summary,
  tag,
}: DetailsProps) => {
  const ref = useRef<HTMLElement>(null);

  const { onMounted } = useMounted();
  const [isOpened, setOpened] = useToggle(open);

  useEffect(() => {
    if (typeof hasNativeDetails === 'undefined') {
      hasNativeDetails =
        typeof document.createElement('details').open === 'boolean';
    }
  }, []);

  useEffect(() => {
    setOpened(open);
  }, [open, setOpened]);

  useEffect(() => {
    if (hasNativeDetails) {
      return;
    }

    if (ref.current) {
      // Don't really need to include this, but just
      // polyfills DOM behaviour for IE/Edge
      if (isOpened) {
        ref.current.setAttribute('open', '');
      } else {
        ref.current.removeAttribute('open');
      }
    }
  }, [isOpened]);

  return (
    <details
      className={classNames('govuk-details', className)}
      open={open}
      ref={ref}
      role={onMounted('group')}
    >
      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */}
      <summary
        aria-controls={onMounted(id)}
        aria-expanded={onMounted(isOpened)}
        className="govuk-details__summary"
        role={onMounted('button')}
        tabIndex={onMounted(0)}
        onClick={event => {
          event.persist();

          if (onToggle) {
            onToggle(!isOpened, event);
            if (analytics) {
              analytics({
                category: 'Statistics summary',
                action: `${summary} - ${tag}`,
              });
            }

            if (event.isDefaultPrevented()) {
              return;
            }
          }

          setOpened(!isOpened);
        }}
        onKeyPress={event => {
          if (event.key === 'Enter' || event.key === ' ') {
            event.preventDefault();
            (event.target as HTMLElement).click();
          }
        }}
        onKeyUp={event => {
          if (event.key === ' ') {
            event.preventDefault();
          }
        }}
      >
        <span
          className="govuk-details__summary-text"
          data-testid="details--expand"
        >
          {summary}
          {tag && (
            <span className="govuk-tag govuk-!-margin-left-2">{tag}</span>
          )}
        </span>
      </summary>
      <div
        aria-hidden={onMounted(!isOpened)}
        className="govuk-details__text"
        id={onMounted(id)}
        style={onMounted(
          !hasNativeDetails
            ? {
                display: !isOpened ? 'none' : undefined,
              }
            : undefined,
        )}
      >
        {children}
      </div>
    </details>
  );
};

export default Details;

export const openAllParentDetails = (target: HTMLElement) => {
  const textContainers = findAllParents(target, '.govuk-details__text');

  textContainers.forEach(textContainer => {
    const summary = document.querySelector<HTMLElement>(
      `summary[aria-controls="${textContainer.id}"]`,
    );

    if (summary && summary.getAttribute('aria-expanded') === 'false') {
      summary.click();
    }
  });
};
