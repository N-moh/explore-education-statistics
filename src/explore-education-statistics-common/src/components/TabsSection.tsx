import { useDesktopMedia } from '@common/hooks/useMedia';
import classNames from 'classnames';
import React, {
  createElement,
  forwardRef,
  HTMLAttributes,
  ReactNode,
} from 'react';
import styles from './TabsSection.module.scss';

export interface TabsSectionProps {
  children: ReactNode;
  id: string;
  /**
   * Set to true if children should not be
   * rendered until tab has been selected.
   */
  lazy?: boolean;
  title: string;
  headingTag?: 'h2' | 'h3' | 'h4';
}

const TabsSection = forwardRef<HTMLElement, TabsSectionProps>(
  (
    { children, id, title, headingTag = 'h3', ...restProps }: TabsSectionProps,
    ref,
  ) => {
    const { onMedia } = useDesktopMedia();

    // Hide additional props from the component's public API to
    // avoid any confusion over this component's usage as
    // it should only be used in combination with Tabs.
    const tabProps = restProps as HTMLAttributes<HTMLElement>;

    return (
      <section
        aria-labelledby={onMedia(tabProps['aria-labelledby'])}
        className={classNames(
          'govuk-tabs__panel',
          'dfe-content-overflow',
          styles.panel,
          {
            'govuk-tabs__panel--hidden': tabProps.hidden,
          },
        )}
        id={id}
        ref={ref}
        role={onMedia('tabpanel')}
        tabIndex={onMedia(-1)}
        data-testid={title}
      >
        {createElement(headingTag, { children: title })}
        {children}
      </section>
    );
  },
);

TabsSection.displayName = 'TabsSection';

export default TabsSection;
