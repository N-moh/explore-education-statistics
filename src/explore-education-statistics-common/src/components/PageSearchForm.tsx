import {
  accordionSectionClasses,
  openAllParentAccordionSections,
} from '@common/components/AccordionSection';
import { openAllParentDetails } from '@common/components/Details';
import FormComboBox from '@common/components/form/FormComboBox';
import { openAllParentTabSections } from '@common/components/TabsSection';
import findAllByText from '@common/lib/dom/findAllByText';
import findAllParents from '@common/lib/dom/findAllParents';
import findPreviousSibling from '@common/lib/dom/findPreviousSibling';
import classNames from 'classnames';
import debounce from 'lodash/debounce';
import React, { Component, ReactNode } from 'react';
import Highlighter from 'react-highlight-words';
import styles from './PageSearchForm.module.scss';

interface SearchResult {
  element: Element;
  scrollIntoView: () => void;
  text: ReactNode;
  location: string;
}

interface Props {
  className?: string;
  elementSelectors: string[];
  id: string;
  minInput: number;
}

interface State {
  searchResults: SearchResult[];
  searchComplete: boolean;
}

class PageSearchForm extends Component<Props, State> {
  public state: State = {
    searchResults: [],
    searchComplete: false,
  };

  public static defaultProps = {
    elementSelectors: ['p', 'li > strong', 'h2', 'h3', 'h4'],
    id: 'pageSearchForm',
    minInput: 3,
  };

  private static getLocationText(element: HTMLElement): string {
    const location: string[] = [];

    const prependNearestHeading = (currentElement: HTMLElement) => {
      const elementTag = currentElement.tagName.toLowerCase();

      if (['h2', 'h3', 'h4'].includes(elementTag)) {
        return;
      }

      const headingEl = findPreviousSibling(currentElement, `h2, h3, h4`);

      if (headingEl) {
        location.unshift(headingEl.textContent || '');
      }
    };

    prependNearestHeading(element);

    findAllParents(element, '*').forEach(parent => {
      prependNearestHeading(parent);

      if (
        parent.parentElement &&
        parent.parentElement.classList.contains(accordionSectionClasses.section)
      ) {
        const accordionHeader = parent.parentElement.querySelector(
          `.${accordionSectionClasses.sectionButton}`,
        );

        if (accordionHeader) {
          location.unshift(accordionHeader.textContent || '');
        }
      }
    });

    return location.join(' > ');
  }

  private search = (value: string) => {
    const { elementSelectors, minInput } = this.props;

    if (value.length < minInput) {
      return;
    }

    const elements = findAllByText(value, elementSelectors.join(', '));

    const searchResults = elements.map(element => {
      const location = PageSearchForm.getLocationText(element);

      return {
        element,
        location,
        text: (
          <Highlighter
            searchWords={[value]}
            textToHighlight={element.textContent || ''}
          />
        ),
        scrollIntoView: () => {
          openAllParentAccordionSections(element);
          openAllParentTabSections(element);
          openAllParentDetails(element);

          this.resetSearch();

          setTimeout(() => {
            // Bit of a hack, but hopefully screen readers will
            // still change focus to the selected element even if we
            // proceed to remove the tabindex shortly afterwards
            // TODO: Verify this works
            const previousTabIndex = element.getAttribute('tabindex');

            if (!previousTabIndex) {
              element.setAttribute('tabindex', '-1');
            }

            element.focus();
            element.scrollIntoView({
              behavior: 'smooth',
              block: 'start',
            });

            if (previousTabIndex) {
              element.setAttribute('tabindex', previousTabIndex);
            } else {
              element.removeAttribute('tabindex');
            }
          });
        },
      };
    });

    this.setState({ searchResults, searchComplete: true });
  };

  // eslint-disable-next-line react/sort-comp
  private debouncedSearch = debounce(this.search, 1000);

  private resetSearch() {
    this.setState({ searchResults: [], searchComplete: false });
  }

  public render() {
    const { className, id } = this.props;
    const { searchResults, searchComplete } = this.state;

    return (
      <form
        className={classNames(styles.container, className)}
        onSubmit={e => e.preventDefault()}
        autoComplete="off"
        role="search"
      >
        <FormComboBox
          classes={{
            inputLabel: 'govuk-visually-hidden',
          }}
          id={id}
          inputProps={{
            placeholder: 'Search this page',
          }}
          inputLabel="Find on this page"
          afterInput={({ value }) => (
            <button
              type="submit"
              className={styles.searchButton}
              value="Search this page"
              onClick={() => this.search(value)}
            />
          )}
          listBoxLabelId={`${id}-resultsLabel`}
          listBoxLabel={() => (
            <div id={`${id}-resultsLabel`} className={styles.resultsLabel}>
              Found <strong>{searchResults.length}</strong>
              {` ${searchResults.length === 1 ? 'result' : 'results'}`}
            </div>
          )}
          options={
            searchComplete
              ? searchResults.map(result => {
                  return (
                    <>
                      <div className={styles.resultHeader}>{result.text}</div>
                      {result.location && (
                        <div className={styles.resultLocation}>
                          {result.location}
                        </div>
                      )}
                    </>
                  );
                })
              : undefined
          }
          onInputChange={event => {
            this.debouncedSearch(event.target.value);
          }}
          onSelect={selectedItem => {
            if (searchResults[selectedItem]) {
              searchResults[selectedItem].scrollIntoView();
            }
          }}
        />
      </form>
    );
  }
}

export default PageSearchForm;
