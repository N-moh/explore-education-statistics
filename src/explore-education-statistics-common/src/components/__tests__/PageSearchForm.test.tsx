import Accordion from '@common/components/Accordion';
import AccordionSection from '@common/components/AccordionSection';
import Details from '@common/components/Details';
import Tabs from '@common/components/Tabs';
import TabsSection from '@common/components/TabsSection';
import React from 'react';
import { fireEvent, render, wait } from 'react-testing-library';
import PageSearchForm from '../PageSearchForm';

describe('PageSearchForm', () => {
  test('renders results found in default elements', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <PageSearchForm />

        <div>
          <p>Not me</p>
          <p>Test 1</p>
        </div>

        <div>
          <h2>Not me</h2>
          <h2>Test 2</h2>
        </div>

        <div>
          <h3>Not me</h3>
          <h3>Test 3</h3>
        </div>

        <div>
          <h4>Not me</h4>
          <h4>Test 4</h4>
        </div>

        <ul>
          <li>Not me</li>
          <li>
            <strong>Test 5</strong>
          </li>
        </ul>
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    expect(
      container.querySelector('#pageSearchForm-resultsLabel'),
    ).toHaveTextContent('Found 5 results');

    const options = container.querySelectorAll('[role="option"]');

    expect(options).toHaveLength(5);
    expect(options[0]).toHaveTextContent('Test 1');
    expect(options[1]).toHaveTextContent('Test 2');
    expect(options[2]).toHaveTextContent('Test 3');
    expect(options[3]).toHaveTextContent('Test 4');
    expect(options[4]).toHaveTextContent('Test 5');

    expect(container.querySelector('[role="listbox"]')).toMatchSnapshot();
  });

  test('renders results found in custom selectors', () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <PageSearchForm elementSelectors={['.target-1', '.target-2 p']} />

        <div>
          <p>Test</p>
        </div>
        <div>
          <h2>Test</h2>
        </div>
        <div>
          <h3>Test</h3>
        </div>
        <div>
          <h4>Test</h4>
        </div>
        <ul>
          <li>
            <strong>Test</strong>
          </li>
        </ul>

        <div className="target-1">
          <p>Test 1</p>
        </div>

        <div className="target-2">
          <h2>Section 1</h2>

          <p>Test 2</p>
        </div>
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    expect(
      container.querySelector('#pageSearchForm-resultsLabel'),
    ).toHaveTextContent('Found 2 results');

    const options = container.querySelectorAll('[role="option"]');

    expect(options).toHaveLength(2);
    expect(options[0]).toHaveTextContent('Test 1');
    expect(options[1]).toHaveTextContent('Test 2');

    expect(container.querySelector('[role="listbox"]')).toMatchSnapshot();
  });

  test('result locations uses nearest heading', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <PageSearchForm />

        <div>
          <h2>Section 1</h2>
          <p>Test 1</p>
        </div>

        <h2>Not this section</h2>
        <h3>Section 2</h3>

        <ul>
          <li>
            <strong>Test 2</strong>
          </li>
        </ul>
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    expect(
      container.querySelector('#pageSearchForm-resultsLabel'),
    ).toHaveTextContent('Found 2 results');

    const options = container.querySelectorAll('[role="option"]');

    expect(options).toHaveLength(2);
    expect(options[0]).toHaveTextContent('Test 1');
    expect(options[0]).toHaveTextContent('Section 1');
    expect(options[1]).toHaveTextContent('Test 2');
    expect(options[1]).toHaveTextContent('Section 2');

    expect(container.querySelector('[role="listbox"]')).toMatchSnapshot();
  });

  test('result locations include heading hierarchy', () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <PageSearchForm />

        <h2>Section 1</h2>

        <div>
          <h3>Section 2</h3>

          <div>
            <h4>Section 3</h4>
            <p>Test 1</p>
          </div>
        </div>
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    expect(
      container.querySelector('#pageSearchForm-resultsLabel'),
    ).toHaveTextContent('Found 1 result');

    const options = container.querySelectorAll('[role="option"]');

    expect(options).toHaveLength(1);
    expect(options[0]).toHaveTextContent('Test 1');
    expect(options[0]).toHaveTextContent('Section 1 > Section 2 > Section 3');

    expect(container.querySelector('[role="listbox"]')).toMatchSnapshot();
  });

  test('result locations include accordion sections', () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <PageSearchForm />

        <h2>Section 1</h2>

        <div>
          <h3>Section 2</h3>

          <Accordion id="test-accordion">
            <AccordionSection heading="Accordion section 1">
              <h4>Section 3</h4>

              <p>Test 1</p>
            </AccordionSection>
            <AccordionSection heading="Accordion section 2">
              <p>Test 2</p>
            </AccordionSection>
          </Accordion>
        </div>
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    expect(
      container.querySelector('#pageSearchForm-resultsLabel'),
    ).toHaveTextContent('Found 2 results');

    const options = container.querySelectorAll('[role="option"]');

    expect(options).toHaveLength(2);
    expect(options[0]).toHaveTextContent('Test 1');
    expect(options[0]).toHaveTextContent(
      'Section 1 > Section 2 > Accordion section 1 > Section 3',
    );
    expect(options[1]).toHaveTextContent('Test 2');
    expect(options[1]).toHaveTextContent(
      'Section 1 > Section 2 > Accordion section 2',
    );

    expect(container.querySelector('[role="listbox"]')).toMatchSnapshot();
  });

  test('clicking result scrolls and focuses the element', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <h2>Section 1</h2>
        <p id="target">Test</p>

        <PageSearchForm />
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    fireEvent.click(container.querySelector('[role="option"]') as HTMLElement);

    jest.runOnlyPendingTimers();

    const target = container.querySelector('#target');

    expect(target).toHaveFocus();
    expect(target).toHaveScrolledIntoView();
  });

  test('pressing Enter on result scrolls and focuses the element', () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <h2>Section 1</h2>
        <p id="target">Test</p>

        <PageSearchForm />
      </div>,
    );

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    const listBox = container.querySelector('[role="listbox"]') as HTMLElement;

    fireEvent.keyDown(listBox, { key: 'ArrowDown' });
    fireEvent.keyDown(listBox, { key: 'Enter' });

    jest.runOnlyPendingTimers();

    const target = container.querySelector('#target');

    expect(target).toHaveFocus();
    expect(target).toHaveScrolledIntoView();
  });

  test('opens parent accordion of selected result', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <Accordion id="test-accordion">
          <AccordionSection heading="Section 1">
            <p id="target">Test</p>
          </AccordionSection>
        </Accordion>

        <PageSearchForm />
      </div>,
    );

    await wait();

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    const accordionSection = container.querySelector(
      '#test-accordion-heading-1',
    );

    expect(accordionSection).toHaveAttribute('aria-expanded', 'false');

    fireEvent.click(container.querySelector('[role="option"]') as HTMLElement);

    jest.runOnlyPendingTimers();

    expect(accordionSection).toHaveAttribute('aria-expanded', 'true');
  });

  test('opens parent details of selected result', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <Details summary="Details 1">
          <p id="target">Test</p>
        </Details>

        <PageSearchForm />
      </div>,
    );

    await wait();

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    const summary = container.querySelector('summary');

    expect(summary).toHaveAttribute('aria-expanded', 'false');

    fireEvent.click(container.querySelector('[role="option"]') as HTMLElement);

    jest.runOnlyPendingTimers();

    expect(summary).toHaveAttribute('aria-expanded', 'true');
  });

  test('opens parent tab section of selected result', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <Tabs>
          <TabsSection title="Tab 1" id="section-1">
            <p>Something</p>
          </TabsSection>
          <TabsSection title="Tab 2" id="section-2">
            <p id="target">Test</p>
          </TabsSection>
        </Tabs>

        <PageSearchForm />
      </div>,
    );

    await wait();

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    const tabs = container.querySelectorAll('[role="tab"]');

    expect(tabs[0]).toHaveTextContent('Tab 1');
    expect(tabs[1]).toHaveTextContent('Tab 2');

    expect(tabs[0]).toHaveAttribute('aria-selected', 'true');
    expect(tabs[1]).toHaveAttribute('aria-selected', 'false');

    fireEvent.click(container.querySelector('[role="option"]') as HTMLElement);

    jest.runOnlyPendingTimers();

    expect(tabs[0]).toHaveAttribute('aria-selected', 'false');
    expect(tabs[1]).toHaveAttribute('aria-selected', 'true');
  });

  test('opens nested sections of selected result', async () => {
    jest.useFakeTimers();

    const { container, getByLabelText } = render(
      <div>
        <Accordion id="test-accordion">
          <AccordionSection heading="Section 1">
            <Tabs>
              <TabsSection title="Tab 1" id="section-1">
                <p>Something</p>
              </TabsSection>
              <TabsSection title="Tab 2" id="section-2">
                <Details summary="Details 1">
                  <p id="target">Test</p>
                </Details>
              </TabsSection>
            </Tabs>
          </AccordionSection>
        </Accordion>

        <PageSearchForm />
      </div>,
    );

    await wait();

    fireEvent.change(getByLabelText('Find on this page'), {
      target: {
        value: 'Test',
      },
    });

    jest.runOnlyPendingTimers();

    const accordionSection = container.querySelector(
      '#test-accordion-heading-1',
    );
    const tabs = container.querySelectorAll('[role="tab"]');
    const summary = container.querySelector('summary');

    fireEvent.click(container.querySelector('[role="option"]') as HTMLElement);

    jest.runOnlyPendingTimers();

    expect(accordionSection).toHaveAttribute('aria-expanded', 'true');
    expect(tabs[1]).toHaveAttribute('aria-selected', 'true');
    expect(summary).toHaveAttribute('aria-expanded', 'true');
  });
});
