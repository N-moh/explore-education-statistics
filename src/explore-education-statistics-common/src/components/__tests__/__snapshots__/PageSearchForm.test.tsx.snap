// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSearchForm renders results found in custom selectors 1`] = `
<ul
  aria-labelledby="pageSearchForm-resultsLabel"
  class="options"
  id="pageSearchForm-options"
  role="listbox"
  tabindex="-1"
>
  <li
    aria-selected="false"
    id="pageSearchForm-option-0"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           1
        </span>
      </span>
    </div>
    
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-1"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           2
        </span>
      </span>
    </div>
    <div
      class="resultLocation"
    >
      Section 1
    </div>
  </li>
</ul>
`;

exports[`PageSearchForm renders results found in default elements 1`] = `
<ul
  aria-labelledby="pageSearchForm-resultsLabel"
  class="options"
  id="pageSearchForm-options"
  role="listbox"
  tabindex="-1"
>
  <li
    aria-selected="false"
    id="pageSearchForm-option-0"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           1
        </span>
      </span>
    </div>
    
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-1"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           2
        </span>
      </span>
    </div>
    
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-2"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           3
        </span>
      </span>
    </div>
    
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-3"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           4
        </span>
      </span>
    </div>
    
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-4"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           5
        </span>
      </span>
    </div>
    
  </li>
</ul>
`;

exports[`PageSearchForm result locations include accordion sections 1`] = `
<ul
  aria-labelledby="pageSearchForm-resultsLabel"
  class="options"
  id="pageSearchForm-options"
  role="listbox"
  tabindex="-1"
>
  <li
    aria-selected="false"
    id="pageSearchForm-option-0"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           1
        </span>
      </span>
    </div>
    <div
      class="resultLocation"
    >
      Section 1 &gt; Section 2 &gt; Accordion section 1 &gt; Section 3
    </div>
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-1"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           2
        </span>
      </span>
    </div>
    <div
      class="resultLocation"
    >
      Section 1 &gt; Section 2 &gt; Accordion section 2
    </div>
  </li>
</ul>
`;

exports[`PageSearchForm result locations include heading hierarchy 1`] = `
<ul
  aria-labelledby="pageSearchForm-resultsLabel"
  class="options"
  id="pageSearchForm-options"
  role="listbox"
  tabindex="-1"
>
  <li
    aria-selected="false"
    id="pageSearchForm-option-0"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           1
        </span>
      </span>
    </div>
    <div
      class="resultLocation"
    >
      Section 1 &gt; Section 2 &gt; Section 3
    </div>
  </li>
</ul>
`;

exports[`PageSearchForm result locations uses nearest heading 1`] = `
<ul
  aria-labelledby="pageSearchForm-resultsLabel"
  class="options"
  id="pageSearchForm-options"
  role="listbox"
  tabindex="-1"
>
  <li
    aria-selected="false"
    id="pageSearchForm-option-0"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           1
        </span>
      </span>
    </div>
    <div
      class="resultLocation"
    >
      Section 1
    </div>
  </li>
  <li
    aria-selected="false"
    id="pageSearchForm-option-1"
    role="option"
  >
    <div
      class="resultHeader"
    >
      <span>
        <mark
          class=" "
        >
          Test
        </mark>
        <span
          class=""
        >
           2
        </span>
      </span>
    </div>
    <div
      class="resultLocation"
    >
      Section 2
    </div>
  </li>
</ul>
`;
